File: app/src/main/java/com/example/mutidemo/ui/GridViewActivity.java
package com.example.mutidemo.ui;

import android.view.View;

import androidx.recyclerview.widget.GridLayoutManager;

import com.example.mutidemo.databinding.ActivityGridviewBinding;
import com.example.mutidemo.util.GlideLoadEngine;
import com.luck.picture.lib.basic.PictureSelector;
import com.luck.picture.lib.config.SelectMimeType;
import com.luck.picture.lib.entity.LocalMedia;
import com.luck.picture.lib.interfaces.OnResultCallbackListener;
import com.pengxh.androidx.lite.adapter.EditableImageAdapter;
import com.pengxh.androidx.lite.base.AndroidxBaseActivity;
import com.pengxh.androidx.lite.utils.ImageUtil;

import java.util.ArrayList;
import java.util.List;

public class GridViewActivity extends AndroidxBaseActivity<ActivityGridviewBinding> {

    private EditableImageAdapter editableImageAdapter;
    private final ArrayList<String> recyclerViewImages = new ArrayList<>();

    @Override
    protected void setupTopBarLayout() {

    }

    @Override
    public void initData() {
        editableImageAdapter = new EditableImageAdapter(this, 9,1f);
        viewBinding.nineRecyclerView.setAdapter(editableImageAdapter);
        editableImageAdapter.setOnItemClickListener(new EditableImageAdapter.OnItemClickListener() {
            @Override
            public void onAddImageClick() {
                selectPicture();
            }

            @Override
            public void onItemClick(int position) {
                ImageUtil.showBigImage(GridViewActivity.this, position, recyclerViewImages);
            }

            @Override
            public void onItemLongClick(View view, int position) {
                editableImageAdapter.deleteImage(position);
            }
        });
    }

    private void selectPicture() {
        PictureSelector.create(GridViewActivity.this)
                .openGallery(SelectMimeType.ofImage())
                .isGif(false)
                .isMaxSelectEnabledMask(true)
                .setFilterMinFileSize(100)
                .setMaxSelectNum(9)
                .isDisplayCamera(false)
                .setImageEngine(GlideLoadEngine.getInstance())
                .forResult(new OnResultCallbackListener<LocalMedia>() {
                    @Override
                    public void onResult(ArrayList<LocalMedia> result) {
                        for (LocalMedia media : result) {
                            recyclerViewImages.add(media.getRealPath());
                        }
                        editableImageAdapter.setupImage(recyclerViewImages);
                    }

                    @Override
                    public void onCancel() {

                    }
                });
    }

    @Override
    public void initEvent() {
        viewBinding.button.setOnClickListener(v -> viewBinding.imagePathView.setText(reformatURL(recyclerViewImages)));
    }

    private String reformatURL(List<String> urls) {
        if (urls.isEmpty()) {
            return "";
        }
        StringBuilder builder = new StringBuilder();
        for (int i = 0; i < urls.size(); i++) {
            if (i != urls.size() - 1) {
                builder.append(urls.get(i)).append(",");
            } else {
                builder.append(urls.get(i));
            }
        }
        return builder.toString();
    }
}

File: app/src/main/java/com/example/multidemo/view/GridViewActivity.kt
package com.example.multidemo.view

import android.os.Bundle
import android.os.Handler
import android.os.Message
import android.view.View
import androidx.lifecycle.lifecycleScope
import com.example.multidemo.R
import com.example.multidemo.databinding.ActivityGridviewBinding
import com.example.multidemo.extensions.initImmersionBar
import com.example.multidemo.util.GlideLoadEngine
import com.luck.picture.lib.basic.PictureSelector
import com.luck.picture.lib.config.SelectMimeType
import com.luck.picture.lib.entity.LocalMedia
import com.luck.picture.lib.interfaces.OnResultCallbackListener
import com.pengxh.kt.lite.adapter.EditableImageAdapter
import com.pengxh.kt.lite.base.KotlinBaseActivity
import com.pengxh.kt.lite.divider.RecyclerViewItemOffsets
import com.pengxh.kt.lite.extensions.createCompressImageDir
import com.pengxh.kt.lite.extensions.dp2px
import com.pengxh.kt.lite.extensions.getScreenWidth
import com.pengxh.kt.lite.extensions.navigatePageTo
import com.pengxh.kt.lite.utils.WeakReferenceHandler
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch
import top.zibin.luban.Luban
import top.zibin.luban.OnCompressListener
import java.io.File

class GridViewActivity : KotlinBaseActivity<ActivityGridviewBinding>(), Handler.Callback {

    private val context = this
    private val recyclerViewImages = ArrayList<String>()
    private val selectedImages = ArrayList<LocalMedia>()
    private val marginOffset by lazy { 1.dp2px(this) }
    private val weakReferenceHandler by lazy { WeakReferenceHandler(this) }
    private lateinit var imageAdapter: EditableImageAdapter

    override fun setupTopBarLayout() {
        binding.rootView.initImmersionBar(this, true, R.color.white)
    }

    override fun initViewBinding(): ActivityGridviewBinding {
        return ActivityGridviewBinding.inflate(layoutInflater)
    }

    override fun observeRequestState() {

    }

    override fun initOnCreate(savedInstanceState: Bundle?) {
        imageAdapter = EditableImageAdapter(this, recyclerViewImages, getScreenWidth(), 9, 3)
        binding.recyclerView.addItemDecoration(
            RecyclerViewItemOffsets(marginOffset, marginOffset, marginOffset, marginOffset)
        )
        binding.recyclerView.adapter = imageAdapter
        imageAdapter.setOnItemClickListener(object : EditableImageAdapter.OnItemClickListener {
            override fun onAddImageClick() {
                selectPicture()
            }

            override fun onItemClick(position: Int) {
                navigatePageTo<BigImageActivity>(position, recyclerViewImages)
            }

            override fun onItemLongClick(view: View?, position: Int) {
                selectedImages.removeAt(position)
                recyclerViewImages.removeAt(position)
                imageAdapter.notifyDataSetChanged()
            }
        })
    }

    override fun handleMessage(msg: Message): Boolean {
        if (msg.what == 2024042301) {
            val file = msg.obj as File

            recyclerViewImages.add(file.absolutePath)
            imageAdapter.notifyDataSetChanged()
        }
        return true
    }

    private fun selectPicture() {
        PictureSelector.create(this).openGallery(SelectMimeType.ofImage()).isGif(false)
            .isMaxSelectEnabledMask(true).setFilterMinFileSize(100).setMaxSelectNum(9)
            .isDisplayCamera(false).setImageEngine(GlideLoadEngine.get)
            .setSelectedData(selectedImages)
            .forResult(object : OnResultCallbackListener<LocalMedia> {
                override fun onResult(result: ArrayList<LocalMedia>) {
                    //因为设置了selectedImages，每次选择数据都会发生变化，所以需要清空之前的缓存
                    selectedImages.clear()
                    recyclerViewImages.clear()

                    //数据链处理已选的图片
                    lifecycleScope.launch {
                        flow {
                            result.forEach {
                                emit(it)
                                delay(1000)
                            }
                        }.collect {
                            selectedImages.add(it)
                            //模拟压缩图片并上传
                            Luban.with(context).load(it.realPath).ignoreBy(100)
                                .setTargetDir(context.createCompressImageDir().toString())
                                .setCompressListener(object : OnCompressListener {
                                    override fun onStart() {

                                    }

                                    override fun onSuccess(file: File) {
                                        //模拟上传图片
                                        val message = weakReferenceHandler.obtainMessage()
                                        message.what = 2024042301
                                        message.obj = file
                                        weakReferenceHandler.sendMessageDelayed(message, 500)
                                    }

                                    override fun onError(e: Throwable) {
                                        e.printStackTrace()
                                    }
                                }).launch()
                        }
                    }
                }

                override fun onCancel() {}
            })
    }

    override fun initEvent() {

    }
}